dbms/src/DataStreams/RuntimeFilter.cpp

```diff
@@ -0,0 +1,140 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "RuntimeFilter.h"
+#include <common/logger_useful.h>
+#include <Interpreters/Set.h>
+#include <Storages/DeltaMerge/FilterParser/FilterParser.h>
+
+namespace DB {
+
+namespace ErrorCodes {
+extern const int SET_SIZE_LIMIT_EXCEEDED;
+}
+
+std::string RuntimeFilter::getSourceColumnName() {
+    return source_column_name;
+}
+
+tipb::Expr RuntimeFilter::getRFExpr() {
+    return rf_expr;
+}
+
+tipb::RuntimeFilterType RuntimeFilter::getRFType() {
+    return rf_type;
+}
+
+tipb::Expr RuntimeFilter::getSourceExpr() {
+    return source_expr;
+}
+
+std::string RuntimeFilter::getFailedReason() {
+    return failed_reason;
+}
+
+void RuntimeFilter::setSourceColumnName(std::string & source_column_name_) {
+    this->source_column_name = source_column_name_;
+}
+
+void RuntimeFilter::setINValuesSet(std::shared_ptr<Set> in_values_set_) {
+    this->in_values_set = in_values_set_;
+}
+
+// There is a possibility that the same rf is canceled multiple times,
+// but rf does not have thread safety issues.
+void RuntimeFilter::updateValues(const ColumnWithTypeAndName values, const LoggerPtr log) {
+    if (isFailed()) {
+        return;
+    }
+    Block data;
+    data.insert(values);
+    switch (rf_type) {
+        case tipb::IN:
+            try {
+                LOG_DEBUG(log, "update values in rf, data size:" + std::to_string(data.rows()));
+                in_values_set->insertFromBlock(data, true);
+            } catch (Exception &e) {
+                if (e.code() == ErrorCodes::SET_SIZE_LIMIT_EXCEEDED) {
+                    updateStatus(-1, "The rf in values exceed the limit " + std::to_string(RF_MAX_IN_VALUE_SET));
+                    LOG_WARNING(log, "cancel runtime filter id:" + std::to_string(id) +", reason: " + failed_reason);
+                }
+            }
+            break;
+        case tipb::MIN_MAX:
+            // todo
+            break;
+    }
+}
+
+void RuntimeFilter::finalize(const LoggerPtr log) {
```

Comment:
Please always try using references first, to avoid object copy.
```suggestion
void RuntimeFilter::finalize(const LoggerPtr & log) {
```